File: backend/app/api/v1/projects.py
"""
Project API Router — router chỉ lo HTTP, toàn bộ logic ở ProjectService.
"""
from typing import Annotated

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_authenticated_user
from app.models.user import User
from app.schemas.project import (
    AddMemberRequest, MilestoneCreate, MilestoneRead,
    ProjectCreate, ProjectKpiContribution, ProjectListItem,
    ProjectMemberRead, ProjectOverview, ProjectUpdate,
    UpdateMemberRoleRequest, ProjectAnalytics,
)
from app.services.project_service import ProjectService

router = APIRouter()

CurrentUser = Annotated[User, Depends(require_authenticated_user)]
DB = Annotated[Session, Depends(get_db)]


# ═══════════════════════════════════════════════════════════════════════════
# Projects CRUD
# ═══════════════════════════════════════════════════════════════════════════

@router.get("/", response_model=list[ProjectListItem])
def list_projects(
    current_user: CurrentUser, db: DB,
    department_id: int | None = Query(default=None),
    status: str | None       = Query(default=None),
    skip: int                = Query(default=0, ge=0),
    limit: int               = Query(default=50, ge=1, le=200),
):
    return ProjectService(db).list_projects(current_user, department_id, status, skip, limit)


@router.post("/", response_model=ProjectListItem, status_code=status.HTTP_201_CREATED)
def create_project(
    payload: ProjectCreate, current_user: CurrentUser, db: DB,
):
    return ProjectService(db).create_project(payload, current_user)


@router.get("/dashboard", response_model=dict)
def get_dashboard_analytics(current_user: CurrentUser, db: DB):
    return ProjectService(db).get_dashboard_analytics(current_user)


@router.get("/{project_id}", response_model=ProjectOverview)
def get_project_overview(project_id: int, current_user: CurrentUser, db: DB):
    return ProjectService(db).get_project_overview(project_id, current_user)


@router.put("/{project_id}", response_model=ProjectListItem)
def update_project(
    project_id: int, payload: ProjectUpdate,
    current_user: CurrentUser, db: DB,
):
    return ProjectService(db).update_project(project_id, payload, current_user)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_project(project_id: int, current_user: CurrentUser, db: DB):
    ProjectService(db).delete_project(project_id, current_user)


# ═══════════════════════════════════════════════════════════════════════════
# Members
# ═══════════════════════════════════════════════════════════════════════════

@router.post("/{project_id}/members", response_model=ProjectMemberRead,
             status_code=status.HTTP_201_CREATED)
def add_member(
    project_id: int, req: AddMemberRequest,
    current_user: CurrentUser, db: DB,
):
    return ProjectService(db).add_member(project_id, req, current_user)


@router.put("/{project_id}/members/{user_id}", response_model=ProjectMemberRead)
def update_member_role(
    project_id: int, user_id: int,
    req: UpdateMemberRoleRequest,
    current_user: CurrentUser, db: DB,
):
    return ProjectService(db).update_member_role(project_id, user_id, req, current_user)


@router.delete("/{project_id}/members/{user_id}", status_code=status.HTTP_204_NO_CONTENT)
def remove_member(
    project_id: int, user_id: int,
    current_user: CurrentUser, db: DB,
):
    ProjectService(db).remove_member(project_id, user_id, current_user)


# ═══════════════════════════════════════════════════════════════════════════
# Milestones
# ═══════════════════════════════════════════════════════════════════════════

@router.post("/{project_id}/milestones", response_model=MilestoneRead,
             status_code=status.HTTP_201_CREATED)
def create_milestone(
    project_id: int, payload: MilestoneCreate,
    current_user: CurrentUser, db: DB,
):
    return ProjectService(db).create_milestone(project_id, payload, current_user)


@router.patch("/{project_id}/milestones/{milestone_id}/complete",
              response_model=MilestoneRead)
def complete_milestone(
    project_id: int, milestone_id: int,
    current_user: CurrentUser, db: DB,
):
    return ProjectService(db).complete_milestone(project_id, milestone_id, current_user)


# ═══════════════════════════════════════════════════════════════════════════
# Analytics
# ═══════════════════════════════════════════════════════════════════════════

@router.get("/{project_id}/analytics", response_model=ProjectAnalytics)
def get_project_analytics(project_id: int, current_user: CurrentUser, db: DB):
    return ProjectService(db).get_analytics(project_id, current_user)


@router.get("/{project_id}/kpi", response_model=ProjectKpiContribution)
def get_project_kpi(project_id: int, current_user: CurrentUser, db: DB):
    svc = ProjectService(db)
    project = svc._get_project_or_404(project_id)
    svc._check_read_access(project, current_user)
    return svc._build_kpi_contribution(project_id)

File: frontend/src/api/projects.ts
import { apiRequest } from './client';
import { 
  ProjectCreate, ProjectUpdate, ProjectAnalytics, 
  ProjectListItem, DashboardSummary, ProjectOverview,
  ProjectMember, MemberRole, Milestone, KpiContribution, MyProject, TeamWorkload,
  AssignableUser, TaskSummary, ProjectReport,
} from '../types/project';

// Lưu ý: Thêm prefix /api/v1 nếu backend của bạn mount router ở /api/v1
const BASE_PATH = '/projects'; 
const COLLECTION_PATH = `${BASE_PATH}/`;

// ── Projects ──────────────────────────────────────────────────────────────

export const projectApi = {
  // GET: Lấy danh sách
  getAll: (params?: { status?: string; departmentId?: number; managerId?: number; skip?: number; limit?: number }) => {
    const q = new URLSearchParams();
    if (params?.status) q.set('status', params.status);
    if (params?.departmentId) q.set('department_id', String(params.departmentId));
    if (params?.managerId) q.set('manager_id', String(params.managerId));
    if (params?.skip) q.set('skip', String(params.skip));
    if (params?.limit) q.set('limit', String(params.limit));
    const qs = q.toString() ? `?${q.toString()}` : '';
    return apiRequest<ProjectListItem[]>(`${COLLECTION_PATH}${qs}`);
  },

  // GET: Lấy chi tiết 1 dự án
  getById: (id: number) => 
    apiRequest<ProjectOverview>(`${BASE_PATH}/${id}`),

  // POST: Tạo mới
  create: (data: ProjectCreate) => 
    apiRequest<ProjectListItem>(COLLECTION_PATH, {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  // PUT: Cập nhật dự án
  update: (id: number, data: ProjectUpdate) => 
    apiRequest<ProjectListItem>(`${BASE_PATH}/${id}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    }),

  // DELETE: Xóa dự án
  delete: (id: number) => 
    apiRequest<void>(`${BASE_PATH}/${id}`, {
      method: 'DELETE',
    }),

  // GET: Lấy thống kê tiến độ
  getAnalytics: (id: number) => 
    apiRequest<ProjectAnalytics>(`${BASE_PATH}/${id}/analytics`),

  // GET: Dashboard summary
  getDashboard: () =>
    apiRequest<DashboardSummary>(`${BASE_PATH}/dashboard`),

  // GET: KPI contribution
  getKpi: (id: number) =>
    apiRequest<KpiContribution>(`${BASE_PATH}/${id}/kpi`),

  getReport: (id: number) =>
    apiRequest<ProjectReport>(`${BASE_PATH}/${id}/report`),
};

// ── Members ────────────────────────────────────────────────────────────────

export const addMember = (projectId: number, userId: number, role: MemberRole) =>
  apiRequest<ProjectMember>(`${BASE_PATH}/${projectId}/members`, {
    method: 'POST',
    body: JSON.stringify({ user_id: userId, role }),
  });

export const getProjectMembers = (projectId: number) =>
  apiRequest<ProjectMember[]>(`${BASE_PATH}/${projectId}/members`);

export const getAssignableUsers = (projectId: number) =>
  apiRequest<AssignableUser[]>(`${BASE_PATH}/${projectId}/assignable-users`);

export const getProjectTasks = (projectId: number) =>
  apiRequest<TaskSummary[]>(`${BASE_PATH}/${projectId}/tasks`);

export const addProjectMember = (
  projectId: number,
  payload: { user_id: number; role: MemberRole; contribution_share?: number; is_active?: boolean },
) =>
  apiRequest<ProjectMember>(`${BASE_PATH}/${projectId}/members`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });

export const removeMember = (projectId: number, userId: number) =>
  apiRequest<void>(`${BASE_PATH}/${projectId}/members/${userId}`, {
    method: 'DELETE',
  });

export const updateMemberRole = (projectId: number, userId: number, role: MemberRole) =>
  apiRequest<ProjectMember>(`${BASE_PATH}/${projectId}/members/${userId}`, {
    method: 'PUT',
    body: JSON.stringify({ role }),
  });

export const updateProjectMember = (
  projectId: number,
  userId: number,
  payload: { role?: MemberRole; contribution_share?: number; is_active?: boolean },
) =>
  apiRequest<ProjectMember>(`${BASE_PATH}/${projectId}/members/${userId}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  });

export const getMyProjects = () => apiRequest<MyProject[]>('/me/projects');
export const getMyProject = (projectId: number) => apiRequest<MyProject>(`/me/projects/${projectId}`);
export const getMyTasks = (projectId?: number) =>
  apiRequest<TaskSummary[]>(`/me/tasks${projectId ? `?project_id=${projectId}` : ''}`);
export const getManagerProjects = () => apiRequest<ProjectListItem[]>('/manager/projects');
export const getTeamWorkload = () => apiRequest<TeamWorkload[]>('/manager/team-workload');

// ── Milestones ─────────────────────────────────────────────────────────────

export const createMilestone = (projectId: number, data: {
  title: string; description?: string; due_date?: string; weight?: number;
}) =>
  apiRequest<Milestone>(`${BASE_PATH}/${projectId}/milestones`, {
    method: 'POST',
    body: JSON.stringify(data),
  });

export const completeMilestone = (projectId: number, milestoneId: number) =>
  apiRequest<Milestone>(`${BASE_PATH}/${projectId}/milestones/${milestoneId}/complete`, {
    method: 'PATCH',
  });

export const updateMilestone = (
  projectId: number,
  milestoneId: number,
  data: { title?: string; description?: string; due_date?: string; weight?: number },
) =>
  apiRequest<Milestone>(`${BASE_PATH}/${projectId}/milestones/${milestoneId}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  });

export const deleteMilestone = (projectId: number, milestoneId: number) =>
  apiRequest<void>(`${BASE_PATH}/${projectId}/milestones/${milestoneId}`, {
    method: 'DELETE',
  });

// ── Exports for backward compatibility ────────────────────────────────────

export const getProjects = projectApi.getAll;
export const getProjectDashboard = projectApi.getDashboard;
export const getProjectById = projectApi.getById;
export const createProject = projectApi.create;
export const updateProject = projectApi.update;
export const deleteProject = projectApi.delete;
export const getProjectAnalytics = projectApi.getAnalytics;
export const getProjectKpi = projectApi.getKpi;
export const getProjectReport = projectApi.getReport;
